[AArch64] Fix stack probing clobbering flags

momchil-velikov opened the pull request:
Certain stack probing sequences might clobber flags, then we can't use a
block as a prologue if the flags register is a live-in on entry to that
block.

momchil-velikov committed:
[AArch64] Fix stack probing clobbering flags

Certain stack probing sequences might clobber flags, then we can't use a
block as a prologue if the flags register is a live-in on entry to that
block.

yozhu commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -1058,9 +1058,14 @@ bool AArch64FrameLowering::canUseAsPrologue(
 
+  // Certain stack probing sequences might clobber flags, then we can't use
+  // the block as a prologue if the flags register is a live-in.
+  if (TLI->hasInlineStackProbe(*MF) && MBB.isLiveIn(AArch64::NZCV))
```
Do we need to add similar logic for X86 targeting, for which stack probing is also supported?

efriedma-quic replied:
x86 already has a similar check for EFLAGS.  (Not specific to stack probing there, because almost all prologues end up clobbering the flags.)

yozhu replied:
If we check `hasStackProbing()` here, do we need to do the same for the condition fixed in #81878 ?

momchil-velikov committed:
[fixup] Use `hasStackProbing`

The difference now is that the condition would trigger
for Windows, where stack probing does clobber flags.

efriedma-quic reviewed (approved):
LGTM
